src/js/components/Pagination/Pagination.js

```diff
@@ -0,0 +1,207 @@
+  ${props =>
+    props.theme.pagination.container && props.theme.pagination.container.extend}
+`;
+
+const Pagination = forwardRef(
+  (
+    {
+      a11yTitle,
+      numItems,
+      numEdgePages = 1, // number of pages at each edge of page indices
+      // number of page controls in the middle
+      numMiddlePages: numMiddlePagesProp = 3,
+      onChange,
+      page: pageProp,
+      step: stepProp,
+      ...rest
+    },
+    ref,
+  ) => {
+    const theme = useContext(ThemeContext) || defaultProps.theme;
+
+    const [, , defaultPage, step] = usePagination({
+      data: numItems,
+      paginationProps: { page: pageProp, step: stepProp },
+    });
+    const [activePage, setActivePage] = useState(defaultPage);
+
+    const getPageIndices = (begin, end) => {
+      const indices = [];
+      for (let i = begin; i <= end; i += 1) {
+        indices.push(i);
+      }
+      return indices;
+    };
+
+    /* Calculate total number pages */
+    const totalPages = Math.ceil(numItems / step);
+
+    /* Define page indices to display */
+    const beginPages = getPageIndices(1, Math.min(numEdgePages, totalPages));
+    const endPages = getPageIndices(
+      Math.max(totalPages - numEdgePages + 1, numEdgePages + 1),
+      totalPages,
+    );
+
+    let numMiddlePages;
+    if (numMiddlePagesProp < 1) {
+      numMiddlePages = 1;
+      console.warn(
+        // eslint-disable-next-line max-len
+        `Property "numMiddlePages" should not be < 1. One middle page button will be shown. Set "numMiddlePages" >= 1 to remove this warning.`,
+      );
+    } else numMiddlePages = numMiddlePagesProp;
+
+    let startingMiddlePages;
+    // odd
+    if (numMiddlePages % 2)
+      startingMiddlePages = Math.min(
+        activePage - Math.floor(numMiddlePages / 2),
+        totalPages - numEdgePages - numMiddlePages,
+      );
+    // even, cannot split equally around active page
+    // let extra page appear on middlePagesEnd instead
+    else
+      startingMiddlePages = Math.min(
+        activePage - Math.floor(numMiddlePages / 2) + 1,
+        totalPages - numEdgePages - numMiddlePages,
+      );
+
+    const middlePagesBegin = Math.max(startingMiddlePages, numEdgePages + 2);
+    const middlePagesEnd = Math.min(
+      Math.max(
+        activePage + Math.floor(numMiddlePages / 2),
+        numEdgePages + numMiddlePages + 1,
+      ),
+      endPages.length > 0 ? endPages[0] - 2 : totalPages - 1,
+    );
+
+    const middlePages = getPageIndices(middlePagesBegin, middlePagesEnd);
+
+    let beginFlex = [];
+    if (middlePagesBegin > numEdgePages + 2) beginFlex = ['more-prev'];
+    else if (numEdgePages + 1 < totalPages - numEdgePages)
+      beginFlex = [numEdgePages + 1];
+
+    let endFlex = [];
+    if (middlePagesEnd < totalPages - numEdgePages - 1) endFlex = ['more-next'];
+    else if (totalPages - numEdgePages > numEdgePages)
+      endFlex = [totalPages - numEdgePages];
+
+    const navPages = [
```

Comment:
Something is feeling a bit off to use the term "pages" here when it contains things other than pages. Maybe just "controls", since some are to direct pages and other are more generic?